### PR TITLE
Rewrote parser to use ASTNodes.

### PR DESCRIPTION
**Pull request by EraYaN**
This is a rewrite for the output of the parse to output a proper AST.

To keep backwards compatibility we could add an AST to JSON transformation. (Using a NodeTransformer for example)

For now I didn't add it yet, because I don't know if there is a possibility of merging this here anyway. I'd love to hear from you.

### CODE REVIEW

**Comment by klahnakoski** [comment]
@EraYaN  It will take me a little time to go through this.  But first some questions:

* This looks like a part of a bigger plan.  What is the plan? That plan is probably not going to be realized here, but it helps to know what project will be consuming this AST output.  If there is no consumer, then we run the risk of code rot.
* Does this code pass the tests?

File: moz_sql_parser/sql_parser.py [hunk 1]
```diff
@@ -41,281 +46,204 @@ def nothing(*args):
...
+RIGHTJOIN = Keyword("right join", caseless=True).setName("right join").addParseAction(an.RightJoinKeyword.from_tokens).setDebugActions(*debug)
+RIGHTOUTERJOIN = Keyword("right outer join", caseless=True).setName("right outer join").addParseAction(an.RightOuterJoinKeyword.from_tokens).setDebugActions(*debug)
+SELECT = Keyword("select", caseless=True).setName("select").addParseAction(an.SelectKeyword.from_tokens).setDebugActions(*debug)
+THEN = Keyword("then", caseless=True).setName("then").addParseAction(an.ThenKeyword.from_tokens).setDebugActions(*debug)
+UNION = Keyword("union", caseless=True).setName("union").addParseAction(an.UnionKeyword.from_tokens).setDebugActions(*debug)
+UNIONALL = Keyword("union all", caseless=True).setName("union all").addParseAction(an.UnionAllKeyword.from_tokens).setDebugActions(*debug)
+WHEN = Keyword("when", caseless=True).setName("when").addParseAction(an.WhenKeyword.from_tokens).setDebugActions(*debug)
+WHERE = Keyword("where", caseless=True).setName("where").addParseAction(an.WhereKeyword.from_tokens).setDebugActions(*debug)
+WITH = Keyword("with", caseless=True).setName("with").addParseAction(an.WithKeyword.from_tokens).setDebugActions(*debug)
```
Comment (klahnakoski) [inline moz_sql_parser/sql_parser.py]: On one hand I am happy you did this so my linter does not complain about missing variables.  On the other hand it is a wall of redundant code that is difficult to verify.  May I suggest the original `keywords`  be CamelCase, and then perform:

```python
for k in keywords:
    name = k.upper().replace(" ", "")
    locs[name] = value = Keyword(k, caseless=True)
        .setName(k.lower())
        .addParseAction(getattr(an, k+"Keyword").from_tokens)
        .setDebugActions(*debug)
    reserved.append(value)
```

**Comment by EraYaN** [comment]
Right, I'm using this in an SQL to Vivado HLS C++ transpiler project (so for FPGAs). ASTs give much more freedom and are more expressive. For example I can now do constant propagation and the like in python using `NodeTransformers`.

If there is a possibility to have this merged, if I can make a json transform to match the original json format, I might give it a go. I just pushed this PR here to get a feel.

So far I had changed some of the tests to actual AST checking (the ASTNode class supports proper value based equality checking), but since for my project there was no direct need to verify all, I stopped there. But if the json conversion is done correctly and it passes all those tests, that should work for the testing front. 

Right now the print function generates code that is valid python code and when pasted into a python file will generate the same object structure. This is very helpful for making tests.

File: moz_sql_parser/sql_parser.py (hunk 1 above)
Comment (EraYaN) [reply moz_sql_parser/sql_parser.py]: Right, that would work, but as you said PyCharm and other IDEs really do not like it. I'm not happy either way really. I could even add the actual keyword to the ASTNode as a class variable.

Comment (klahnakoski) [reply moz_sql_parser/sql_parser.py]: We can consider putting the keyword generation in its own file, and them imported here.  This way the linters stop complaining about the main code.

**Comment by klahnakoski** [comment]
Re-open if you can get this to pass tests

**Comment by EraYaN** [comment]
Aah yes I doubt I will have time to get this upstream, I wrote this for my thesis to use the parser in an actual compiler. Unless someone needs something like this, I doubt it will see the AST -> JSON converter required to pass existing tests. I might give it a go at some point but it'll be a lot of work.

**Pull request closed without merging**